src/heading.js

```diff
@@ -27,46 +25,120 @@ export default class Heading extends Plugin {
-				id, label
+		const dropdownItems = new Collection();
+		const options = this._getLocalizedOptions();
+		const commands = [];
+		let defaultOption;
+
+		for ( let option of options ) {
+			// Add the option to the collection.
+			dropdownItems.add( new Model( {
+				modelElement: option.modelElement,
+				label: option.title
 			} ) );
+
+			commands.push( editor.commands.get( option.modelElement ) );
+
+			if ( !defaultOption && option.modelElement == 'paragraph' ) {
+				defaultOption = option;
+			}
 		}
 
 		// Create dropdown model.
 		const dropdownModel = new Model( {
 			withText: true,
-			items: collection
+			items: dropdownItems
 		} );
 
-		// Bind dropdown model to command.
-		dropdownModel.bind( 'isEnabled' ).to( command, 'isEnabled' );
-		dropdownModel.bind( 'label' ).to( command, 'value', option => option.label );
+		dropdownModel.bind( 'isEnabled' ).to(
+			// Bind to #isEnabled of each command...
+			...getCommandsBindingTargets( commands, 'isEnabled' ),
+			// ...and set it true if any command #isEnabled is true.
+			( ...areEnabled ) => areEnabled.some( isEnabled => isEnabled )
+		);
+
+		dropdownModel.bind( 'label' ).to(
+			// Bind to #value of each command...
+			...getCommandsBindingTargets( commands, 'value' ),
+			// ...and chose the title of the first one which #value is true.
+			( ...areActive ) => {
+				const index = areActive.findIndex( value => value );
+
+				// If none of the commands is active, display the first one.
+				return ( options[ index ] || defaultOption ).title;
+			}
+		);
 
 		// Register UI component.
 		editor.ui.componentFactory.add( 'headings', ( locale ) => {
 			const dropdown = createListDropdown( dropdownModel, locale );
 
 			// Execute command when an item from the dropdown is selected.
-			this.listenTo( dropdown, 'execute', ( { source: { id } } ) => {
-				editor.execute( 'heading', { id } );
+			this.listenTo( dropdown, 'execute', ( { source: { modelElement } } ) => {
+				editor.execute( modelElement );
 				editor.editing.view.focus();
 			} );
 
 			return dropdown;
 		} );
 	}
+
+	/**
+	 * Returns heading options as defined in `config.heading.options` but processed to consider
+	 * editor localization, i.e. to display {@link module:heading/headingcommand~HeadingOption}
+	 * in the correct language.
+	 *
+	 * Note: The reason behind this method is that there's no way to use {@link module:utils/locale~Locale#t}
+	 * when the user config is defined because the editor does not exist yet.
+	 *
+	 * @private
+	 * @returns {Array.<module:heading/headingcommand~HeadingOption>}.
+	 */
+	_getLocalizedOptions() {
+		const editor = this.editor;
+		const t = editor.t;
+		const localizedTitles = {
+			Paragraph: t( 'Paragraph' ),
+			'Heading 1': t( 'Heading 1' ),
+			'Heading 2': t( 'Heading 2' ),
+			'Heading 3': t( 'Heading 3' )
+		};
+
+		return editor.config.get( 'heading.options' ).map( option => {
+			let title;
+
+			if ( option.modelElement == 'paragraph' ) {
```

Comment:
I meant – why it doesn't have that property? It doesn't make sense that it has `viewElement`. It doesn't also make sense that it has `modelElement` – it could be replaced with `defaultBlock: true` or something (but we can stick to `modelElement` – I'm ok with this). But it certainly can have `title`.